ota_university/ctf_framework/views/challenge.py

```diff
@@ -0,0 +1,70 @@
+from django.contrib.auth.decorators import login_required
+from .base_view import *
+
+
+@login_required()
+def index(request):
+    user = UserProfile.objects.get(user=request.user)
+    challenges = Challenge.objects.all()
+
+    challenge_context = []
+
+    for challenge in challenges:
+        challenge_context.append({"challenge": challenge, "completed": challenge in user.challenges.all()})
+
+    context = {
+        "challenges": challenge_context
+    }
+    return render(request, "challenge/index.html", context)
+
+
+@login_required()
+def show(request, challenge_id):
+    """View the page of a specific challenge."""
+    user = UserProfile.objects.get(user=request.user)
+    message = None
+
+    challenge = Challenge.objects.get(id=challenge_id)
+
+    context = {
+        "challenge": challenge,
+        "user": user,
+        "challenge_completed": challenge in user.challenges.all(),
+        "message": message
+    }
+
+    return render(request, "challenge/show.html", context)
+
+
+@login_required()
+def submit(request, challenge_id):
+    """Submit a flag for a given challenge."""
+
+    user = UserProfile.objects.get(user=request.user)
+    message = None
+
+    challenge = Challenge.objects.get(id=challenge_id)
+
+    if request.method == "POST":
+        flag = request.POST["flag"]
+        try:
+            # Check for matching challenge with this flag
+            challenge = Challenge.objects.get(id=challenge_id, flag=flag)
+
+            # Add this challenge to user's completed challenges
+            user.challenges.add(challenge)
+            user.save()
+            message = "Correct!"
+
+        except ObjectDoesNotExist:
+            message = "Incorrect!"
+
+    context = {
+        "challenge": challenge,
+        "user": user,
+        "challenge_completed": challenge in user.challenges.all(),
+        "message": message
+    }
+
+    return render(request, "challenge/show.html", context)
```

Comment:
Ideally `submit()` only processes the flag submission and does not render pages. We should maybe use `redirect()` to redirect to `show()`.

Reply:
Good idea. I can add a GET parameter for the message response onto the #show and redirect to that.

Reply:
No redirect needed. Instead, it just returns the challenge.show view which now takes a new parameter of Message to display "Correct"/"Incorrect".

Reply:
Yes, a redirect is needed. As I mentioned above, `submit()` should not render pages. It should redirect to another action which renders.

Not to mention that if you keep it as-is, you'll end up having `/submit?message=some_message` in your URL, which is not what we want.